File: ch30/btree.c
#define _GNU_SOURCE
#include "btree.h"

#include <stdlib.h>
#include <string.h>
#include <errno.h>
#include <assert.h>

/* ---- helpers ------------------------------------------------------------ */

static bt_node_t *node_new(const char *key, void *value) {
    bt_node_t *n = calloc(1, sizeof(*n));
    if (!n) return NULL;
    n->key = strdup(key);
    if (!n->key) { free(n); return NULL; }
    n->value = value;
    if (pthread_mutex_init(&n->mtx, NULL) != 0) {
        free(n->key); free(n); return NULL;
    }
    return n;
}

static void node_free(bt_node_t *n, void (*free_value)(void*)) {
    if (!n) return;
    if (free_value) free_value(n->value);
    pthread_mutex_destroy(&n->mtx);
    free(n->key);
    free(n);
}

/* Post-order free; caller must hold tree write-lock. */
static void node_free_recursive(bt_node_t *n, void (*free_value)(void*)) {
    if (!n) return;
    node_free_recursive(n->left, free_value);
    node_free_recursive(n->right, free_value);
    node_free(n, free_value);
}

/* ---- public API --------------------------------------------------------- */

int bt_init(btree_t *t) {
    if (!t) return EINVAL;
    t->root = NULL;
    return pthread_rwlock_init(&t->rwlock, NULL);
}

void bt_destroy(btree_t *t, void (*free_value)(void*)) {
    if (!t) return;
    pthread_rwlock_wrlock(&t->rwlock);
    node_free_recursive(t->root, free_value);
    t->root = NULL;
    pthread_rwlock_unlock(&t->rwlock);
    pthread_rwlock_destroy(&t->rwlock);
}

/* lookup() – readers share the rwlock, no frees can happen concurrently */
bool bt_lookup(btree_t *t, const char *key, void **value_out) {
    if (!t || !key) return false;

    pthread_rwlock_rdlock(&t->rwlock);

    bt_node_t *cur = t->root;
    while (cur) {
        /* Lock this element while we inspect it (as per exercise). */
        pthread_mutex_lock(&cur->mtx);
        int cmp = strcmp(key, cur->key);
        if (cmp == 0) {
            if (value_out) *value_out = cur->value;
            pthread_mutex_unlock(&cur->mtx);
            pthread_rwlock_unlock(&t->rwlock);
            return true;
        }
        bt_node_t *next = (cmp < 0) ? cur->left : cur->right;

        /* Hand-over-hand: lock next before releasing cur?  For a pure reader
           under the tree RD lock it's safe to just drop cur and move on,
           because writers (which could relink) are excluded by the RW lock. */
        pthread_mutex_unlock(&cur->mtx);
        cur = next;
    }

    pthread_rwlock_unlock(&t->rwlock);
    return false;
}

/* add() – writers get the tree write-lock; frees/relinks are exclusive */
int bt_add(btree_t *t, const char *key, void *value) {
    if (!t || !key) return EINVAL;

    int rc = 0;
    pthread_rwlock_wrlock(&t->rwlock);

    if (!t->root) {
        bt_node_t *n = node_new(key, value);
        if (!n) { rc = ENOMEM; goto out; }
        t->root = n;
        goto out;
    }

    bt_node_t *parent = NULL, *cur = t->root;
    pthread_mutex_lock(&cur->mtx);

    for (;;) {
        int cmp = strcmp(key, cur->key);
        if (cmp == 0) {
            /* Replace value; keep key (stable) */
            cur->value = value;
            rc = 1; /* replaced */
            pthread_mutex_unlock(&cur->mtx);
            break;
        }

        bt_node_t **link = (cmp < 0) ? &cur->left : &cur->right;

        if (*link == NULL) {
            /* Insert here */
            bt_node_t *n = node_new(key, value);
            if (!n) { rc = ENOMEM; pthread_mutex_unlock(&cur->mtx); break; }
            *link = n;
            pthread_mutex_unlock(&cur->mtx);
            break;
        }

        /* Move down: lock child, then unlock parent (hand-over-hand) */
        bt_node_t *next = *link;
        pthread_mutex_lock(&next->mtx);
        if (parent) pthread_mutex_unlock(&parent->mtx);
        parent = cur;
        cur = next;
    }

    if (parent) pthread_mutex_unlock(&parent->mtx); /* in case loop exited early */

out:
    pthread_rwlock_unlock(&t->rwlock);
    return rc;
}

/* Helper: find minimum node in a subtree; caller holds write-lock and 'start' locked.
   Returns locked successor node and its parent (parent may be 'start'). */
static void find_min_locked(bt_node_t *start, bt_node_t **min_parent, bt_node_t **min_node) {
    bt_node_t *parent = start;
    bt_node_t *cur = start->right;
    pthread_mutex_lock(&cur->mtx);
    while (cur->left) {
        bt_node_t *next = cur->left;
        pthread_mutex_lock(&next->mtx);
        pthread_mutex_unlock(&parent->mtx);
        parent = cur;
        cur = next;
    }
    *min_parent = parent;
    *min_node = cur;
}

/* delete() – standard BST cases; writer lock excludes readers/writers */
int bt_delete(btree_t *t, const char *key, void **old_value) {
    if (!t || !key) return EINVAL;

    int result = -1;

    pthread_rwlock_wrlock(&t->rwlock);

    bt_node_t *parent = NULL;
    bt_node_t *cur = t->root;
    if (!cur) { result = -1; goto out_unlock; }

    pthread_mutex_lock(&cur->mtx);

    /* Search with hand-over-hand locking */
    while (cur && strcmp(key, cur->key) != 0) {
        int cmp = strcmp(key, cur->key);
        bt_node_t *next = (cmp < 0) ? cur->left : cur->right;
        if (!next) { pthread_mutex_unlock(&cur->mtx); result = -1; goto out_unlock; }
        pthread_mutex_lock(&next->mtx);
        if (parent) pthread_mutex_unlock(&parent->mtx);
        parent = cur;
        cur = next;
    }

    if (!cur) { result = -1; goto out_unlock; } /* not found */

    /* Save value for caller (optional) */
    if (old_value) *old_value = cur->value;

    /* Case 1: node with at most one child */
    if (!cur->left || !cur->right) {
        bt_node_t *child = cur->left ? cur->left : cur->right;
        if (!parent) {
            /* deleting root */
            t->root = child;
        } else {
            /* parent is locked */
            if (parent->left == cur) parent->left = child;
            else parent->right = child;
        }
        pthread_mutex_unlock(&cur->mtx);
        if (parent) pthread_mutex_unlock(&parent->mtx);
        node_free(cur, /*free_value=*/NULL); /* value ownership already handed to caller */
        result = 0;
        goto out_unlock;
    }

    /* Case 2: two children – replace with in-order successor (min in right subtree) */
    bt_node_t *succ_parent = NULL, *succ = NULL;
    find_min_locked(cur, &succ_parent, &succ);
    /* 'cur' and 'succ' and 'succ_parent' are locked; succ has no left child */

    /* Copy successor's key/value into cur (key must remain heap-allocated) */
    char *new_key = strdup(succ->key);
    if (!new_key) {
        /* unwind locks */
        pthread_mutex_unlock(&succ->mtx);
        if (succ_parent) pthread_mutex_unlock(&succ_parent->mtx);
        pthread_mutex_unlock(&cur->mtx);
        if (parent) pthread_mutex_unlock(&parent->mtx);
        result = ENOMEM;
        goto out_unlock;
    }
    free(cur->key);
    cur->key = new_key;
    if (old_value) *old_value = succ->value;  /* hand value of removed node to caller */
    else cur->value = succ->value;            /* or keep successor's value in place */

    /* Splice out successor (succ has no left child) */
    bt_node_t *succ_child = succ->right; /* may be NULL */
    if (succ_parent == cur) {
        cur->right = succ_child;
    } else {
        succ_parent->left = succ_child;
    }

    pthread_mutex_unlock(&succ->mtx);
    pthread_mutex_unlock(&succ_parent->mtx);
    pthread_mutex_unlock(&cur->mtx);
    if (parent) pthread_mutex_unlock(&parent->mtx);

    node_free(succ, /*free_value=*/NULL); /* value ownership already handled */
    result = 0;

out_unlock:
    pthread_rwlock_unlock(&t->rwlock);
    return result;
}

File: ch30/btree.py
from __future__ import annotations
import threading
from typing import Any, Optional, Tuple, Callable


class RWLock:
    """
    Reader–writer lock with writer preference to avoid writer starvation.
    Usage:
        with rw.read_lock(): ...
        with rw.write_lock(): ...
    """
    def __init__(self) -> None:
        self._mu = threading.Lock()
        self._ok_to_read = threading.Condition(self._mu)
        self._ok_to_write = threading.Condition(self._mu)
        self._active_readers = 0
        self._active_writers = 0
        self._waiting_writers = 0

    class _ReadCtx:
        def __init__(self, rw: "RWLock") -> None:
            self.rw = rw

        def __enter__(self):
            rw = self.rw
            with rw._mu:
                # If writers are active or waiting, block new readers
                while rw._active_writers or rw._waiting_writers:
                    rw._ok_to_read.wait()
                rw._active_readers += 1
            return self

        def __exit__(self, exc_type, exc, tb):
            rw = self.rw
            with rw._mu:
                rw._active_readers -= 1
                if rw._active_readers == 0:
                    rw._ok_to_write.notify()
            return False

    class _WriteCtx:
        def __init__(self, rw: "RWLock") -> None:
            self.rw = rw

        def __enter__(self):
            rw = self.rw
            with rw._mu:
                rw._waiting_writers += 1
                while rw._active_writers or rw._active_readers:
                    rw._ok_to_write.wait()
                rw._waiting_writers -= 1
                rw._active_writers = 1
            return self

        def __exit__(self, exc_type, exc, tb):
            rw = self.rw
            with rw._mu:
                rw._active_writers = 0
                # Prefer waking writers; otherwise wake all readers
                if rw._waiting_writers:
                    rw._ok_to_write.notify()
                else:
                    rw._ok_to_read.notify_all()
            return False

    def read_lock(self):  return RWLock._ReadCtx(self)
    def write_lock(self): return RWLock._WriteCtx(self)


class _Node:
    __slots__ = ("key", "value", "left", "right", "lock")

    def __init__(self, key: str, value: Any) -> None:
        self.key = key
        self.value = value
        self.left: Optional["_Node"] = None
        self.right: Optional["_Node"] = None
        self.lock = threading.Lock()


class BTree:
    """
    Thread-safe, unbalanced BST keyed by str with arbitrary values.

    API:
        t = BTree.initialize()
        t.add(key, value)          -> 0 if inserted, 1 if replaced existing
        t.delete(key)              -> (True, old_value) or (False, None)
        t.lookup(key)              -> (True, value) or (False, None)
        t.destroy(free_value=None) -> clear all nodes; optional callback per value

    Locking strategy:
        - Tree-level RWLock:
            * lookup(): read lock (many concurrent readers)
            * add/delete/destroy(): write lock (exclusive)
        - Per-node locks taken while inspecting/mutating that node
          (hand-over-hand style down the path).
    """

    def __init__(self) -> None:
        self._root: Optional[_Node] = None
        self._rw = RWLock()

    @staticmethod
    def initialize() -> "BTree":
        return BTree()

    # ---------------------------- public ops ---------------------------------

    def add(self, key: str, value: Any) -> int:
        """Insert or replace. Returns 0 if inserted, 1 if replaced."""
        with self._rw.write_lock():
            if self._root is None:
                self._root = _Node(key, value)
                return 0

            parent: Optional[_Node] = None
            cur = self._root
            cur.lock.acquire()

            try:
                while True:
                    if key == cur.key:
                        # Replace value
                        cur.value = value
                        return 1
                    go_left = key < cur.key
                    child = cur.left if go_left else cur.right
                    if child is None:
                        # Insert new node here
                        new_node = _Node(key, value)
                        if go_left:
                            cur.left = new_node
                        else:
                            cur.right = new_node
                        return 0
                    # Move down: lock child first, then release parent
                    child.lock.acquire()
                    if parent is not None:
                        parent.lock.release()
                    parent = cur
                    cur = child
            finally:
                # Release any locks still held
                cur.lock.release()
                if parent is not None:
                    parent.lock.release()

    def lookup(self, key: str) -> Tuple[bool, Optional[Any]]:
        """Return (True, value) if found, else (False, None)."""
        with self._rw.read_lock():
            cur = self._root
            while cur is not None:
                cur.lock.acquire()
                try:
                    if key == cur.key:
                        return True, cur.value
                    cur_next = cur.left if key < cur.key else cur.right
                finally:
                    cur.lock.release()
                cur = cur_next
            return False, None

    def delete(self, key: str) -> Tuple[bool, Optional[Any]]:
        """Delete key. Returns (True, old_value) if deleted, else (False, None)."""
        with self._rw.write_lock():
            parent: Optional[_Node] = None
            cur = self._root

            if cur is None:
                return False, None

            cur.lock.acquire()

            # Search down with hand-over-hand locking
            while cur is not None and key != cur.key:
                nxt = cur.left if key < cur.key else cur.right
                if nxt is None:
                    # Not found
                    cur.lock.release()
                    if parent is not None:
                        parent.lock.release()
                    return False, None
                nxt.lock.acquire()
                if parent is not None:
                    parent.lock.release()
                parent = cur
                cur = nxt

            if cur is None:
                if parent is not None:
                    parent.lock.release()
                return False, None

            old_value = cur.value

            # Case 1: at most one child
            if cur.left is None or cur.right is None:
                child = cur.left if cur.left is not None else cur.right
                if parent is None:
                    # Deleting root
                    self._root = child
                    cur.lock.release()
                else:
                    if parent.left is cur:
                        parent.left = child
                    else:
                        parent.right = child
                    # Release in a defined order; parent was locked
                    cur.lock.release()
                    parent.lock.release()
                return True, old_value

            # Case 2: two children — replace with in-order successor
            succ_parent, succ = self._find_min_locked(cur)  # both returned locked

            # Move successor's key/value up to 'cur'
            cur.key, cur.value = succ.key, succ.value

            # Splice out successor (it has no left child)
            succ_child = succ.right
            if succ_parent is cur:
                cur.right = succ_child
            else:
                succ_parent.left = succ_child

            # Release locks carefully to avoid double-release when succ_parent is cur
            succ.lock.release()
            if succ_parent is not cur:
                succ_parent.lock.release()

            # Release cur and (if exists) original search parent
            cur.lock.release()
            if parent is not None:
                parent.lock.release()

            return True, old_value

    def destroy(self, free_value: Optional[Callable[[Any], None]] = None) -> None:
        """Clear the tree (exclusive). Optionally free/dispose each value."""
        def postorder(n: Optional[_Node]) -> None:
            if n is None:
                return
            postorder(n.left)
            postorder(n.right)
            if free_value:
                try:
                    free_value(n.value)
                except Exception:
                    # Best-effort cleanup; ignore callback errors
                    pass

        with self._rw.write_lock():
            postorder(self._root)
            self._root = None

    # ---------------------------- helpers ------------------------------------

    def _find_min_locked(self, start_locked: _Node) -> Tuple[_Node, _Node]:
        """
        Find the in-order successor in start_locked.right subtree.
        Pre: write lock held; start_locked is locked; start_locked.right is not None.
        Returns (parent_locked, min_locked), both still locked on return.
        """
        parent = start_locked
        cur = start_locked.right
        assert cur is not None
        cur.lock.acquire()
        while cur.left is not None:
            nxt = cur.left
            nxt.lock.acquire()
            parent.lock.release()
            parent = cur
            cur = nxt
        return parent, cur


